{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from collections import defaultdict, Counter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Barcode</th>\n",
       "      <th>File</th>\n",
       "      <th>BFAs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAATAAAATTAAAACAGAGTTTGGTCCCATGAACATATTTGTCAAT...</td>\n",
       "      <td>FLC_2N_Plate3_FLC-P3-E8-CGAGGCTG-TACTCCTT_S26</td>\n",
       "      <td>dBFA2;hBFA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AGGAAAATGGCTAATTGTCTTAATGGTTCCCAAGCGAATTATATTT...</td>\n",
       "      <td>CLM_clones_1N_P1-G12-ATCTCAGG-ATTAGACG_S84</td>\n",
       "      <td>hBFA1;hBFA2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             Barcode  \\\n",
       "0  AAATAAAATTAAAACAGAGTTTGGTCCCATGAACATATTTGTCAAT...   \n",
       "1  AGGAAAATGGCTAATTGTCTTAATGGTTCCCAAGCGAATTATATTT...   \n",
       "\n",
       "                                            File         BFAs  \n",
       "0  FLC_2N_Plate3_FLC-P3-E8-CGAGGCTG-TACTCCTT_S26  dBFA2;hBFA2  \n",
       "1     CLM_clones_1N_P1-G12-ATCTCAGG-ATTAGACG_S84  hBFA1;hBFA2  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bcd = pd.read_csv('../BC_extraction/bc_assignments.csv')\n",
    "bcd.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "short_names = {\n",
    "    'ReAnn_Dip_variants.xlsx': 'Dip_clone_Env',\n",
    "    'ReAnn_CLM_R2_1N_variants.xlsx': 'CLM_R2_clones_1N_',\n",
    "    'CLM_2N_Batch2_Filter_InWork_ReAnn.xlsx': 'CLM_2N_Batch2',\n",
    "    'ReAnn_GlyEtOH_variants_InWork.xlsx': 'GlyEtOH_2N_20170616',\n",
    "    'ReAnn_GlyEtOH_1N_variants.xlsx': 'GlyEtOH_1N',\n",
    "    'ReAnn_CLM_variants_InWork.xlsx': 'CLM_2N',\n",
    "    'ReAnn_FLC_R1R2_1N_variants.xlsx': 'FLC_R2_clones_1N',\n",
    "    'ReAnn_FLC_variants_InWork.xlsx': 'FLC_2N_Plate2',\n",
    "    'ReAnn_MiloAddFiles_variants.xlsx': 'GlyEtOH_1N',\n",
    "    'ReAnn_CLM_1N_variants.xlsx': 'CLM_clones_1N'\n",
    "}\n",
    "\n",
    "sheets = {'CLM_2N_Batch2_Filter_InWork_ReAnn.xlsx': 'ReAnn_Manual_Variants',\n",
    "          'CLM_2N_Batch2_SNP_Filter_InWork_short.xlsx': 'ReAnn_Manual_Variants',\n",
    "          'ReAnn_CLM_1N_variants.xlsx': 'Curated_Variants',\n",
    "          'ReAnn_CLM_R2_1N_variants.xlsx': 'Curated_Variants',\n",
    "          'ReAnn_FLC_R1R2_1N_variants.xlsx': 'Curated_Variants',\n",
    "          'ReAnn_FLC_variants_InWork.xlsx': 'Variants'}\n",
    "\n",
    "fds = dict()\n",
    "for f in short_names:\n",
    "    if f in sheets:\n",
    "        fds[f.split('.')[0]] = pd.read_excel('../Mutation_data/'+f, sheet_name=sheets[f])\n",
    "    else:\n",
    "        fds[f.split('.')[0]] = pd.read_excel('../Mutation_data/'+f)\n",
    "    fds[f.split('.')[0]]['Short_name'] = [short_names[f]]*len(fds[f.split('.')[0]])\n",
    "    fds[f.split('.')[0]]['File'] = fds[f.split('.')[0]]['Short_name']+'_'+fds[f.split('.')[0]]['Strain']\n",
    "    \n",
    "        \n",
    "dd = []\n",
    "for f in fds:\n",
    "    if 'Manual.Annotation' in fds[f]:\n",
    "        fds[f] = fds[f].rename(columns={'Manual.Annotation': 'Manual.Verification'})\n",
    "    if 'Manual.annotation' in fds[f]:\n",
    "        fds[f] = fds[f].rename(columns={'Manual.annotation': 'Manual.Verification'})\n",
    "    fds[f]['Run'] = [f]*len(fds[f])\n",
    "    dd.append(fds[f][['Run', 'File', 'GENE', 'EFFECT', 'IMPACT', 'CHROM', 'POS', 'REF', 'ALT', 'Manual.Verification', 'AD', 'DP', 'QUAL']])\n",
    "\n",
    "    \n",
    "def heterozygous_call(ad):\n",
    "    ref, alt = int(ad.split(',')[0]), int(ad.split(',')[1])\n",
    "    tot = ref+alt\n",
    "    if alt < 2:\n",
    "        return np.nan\n",
    "    elif tot < 4:\n",
    "        return np.nan\n",
    "    else:\n",
    "        if alt/tot >= 0.15 and alt/tot <= 0.85:\n",
    "            return 'Heterozygous'\n",
    "        elif alt/tot < 0.15:\n",
    "            return np.nan\n",
    "        else:\n",
    "            return 'Homozygous'\n",
    "\n",
    "muts = pd.concat(dd)\n",
    "muts['Call'] = muts['AD'].apply(heterozygous_call)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Homozygous      1247\n",
       "Heterozygous    1140\n",
       "Name: Call, dtype: int64"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "muts.Call.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CLM_2N_Batch2_Filter_InWork_ReAnn\n",
      "Heterozygous    122\n",
      "Homozygous        2\n",
      "Name: Call, dtype: int64\n",
      "ReAnn_CLM_1N_variants\n",
      "Homozygous      392\n",
      "Heterozygous      6\n",
      "Name: Call, dtype: int64\n",
      "ReAnn_CLM_R2_1N_variants\n",
      "Homozygous      165\n",
      "Heterozygous      3\n",
      "Name: Call, dtype: int64\n",
      "ReAnn_CLM_variants_InWork\n",
      "Heterozygous    73\n",
      "Name: Call, dtype: int64\n",
      "ReAnn_Dip_variants\n",
      "Heterozygous    295\n",
      "Homozygous        2\n",
      "Name: Call, dtype: int64\n",
      "ReAnn_FLC_R1R2_1N_variants\n",
      "Homozygous      442\n",
      "Heterozygous     19\n",
      "Name: Call, dtype: int64\n",
      "ReAnn_FLC_variants_InWork\n",
      "Heterozygous    334\n",
      "Homozygous        4\n",
      "Name: Call, dtype: int64\n",
      "ReAnn_GlyEtOH_1N_variants\n",
      "Homozygous      209\n",
      "Heterozygous     11\n",
      "Name: Call, dtype: int64\n",
      "ReAnn_GlyEtOH_variants_InWork\n",
      "Heterozygous    276\n",
      "Homozygous        7\n",
      "Name: Call, dtype: int64\n",
      "ReAnn_MiloAddFiles_variants\n",
      "Homozygous      24\n",
      "Heterozygous     1\n",
      "Name: Call, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "for run in sorted(set(muts.Run)):\n",
    "    print(run)\n",
    "    print(muts[muts.Run==run].Call.value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Barcode</th>\n",
       "      <th>Genes</th>\n",
       "      <th>mut_IDs</th>\n",
       "      <th>Putative_Drivers</th>\n",
       "      <th>Putative_Driver_Mutations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TGCACAAAATAAAATATCCTTCCTGAGTGTCAATCGATTTTTTTGT...</td>\n",
       "      <td>ACS2;BIO3;GAS3;VPS34</td>\n",
       "      <td>ACS2_XII_446579_T_Heterozygous_NON_SYNONYMOUS_...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TTGTTAAACCAGAATCCTCTTATACAAACGAAACAGTGTTGTAATT...</td>\n",
       "      <td>KTR2</td>\n",
       "      <td>KTR2_XI_558029_C_Heterozygous_NON_SYNONYMOUS_C...</td>\n",
       "      <td>KTR2</td>\n",
       "      <td>KTR2_XI_558029_C_Heterozygous_NON_SYNONYMOUS_C...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             Barcode                 Genes  \\\n",
       "0  TGCACAAAATAAAATATCCTTCCTGAGTGTCAATCGATTTTTTTGT...  ACS2;BIO3;GAS3;VPS34   \n",
       "1  TTGTTAAACCAGAATCCTCTTATACAAACGAAACAGTGTTGTAATT...                  KTR2   \n",
       "\n",
       "                                             mut_IDs Putative_Drivers  \\\n",
       "0  ACS2_XII_446579_T_Heterozygous_NON_SYNONYMOUS_...                    \n",
       "1  KTR2_XI_558029_C_Heterozygous_NON_SYNONYMOUS_C...             KTR2   \n",
       "\n",
       "                           Putative_Driver_Mutations  \n",
       "0                                                     \n",
       "1  KTR2_XI_558029_C_Heterozygous_NON_SYNONYMOUS_C...  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# merging with barcode data also pulled from sequencing files\n",
    "mu = muts.merge(bcd[['File', 'Barcode']], on='File', how='inner')\n",
    "# Filter out synonymous / noncoding mutations and mutations that didn't pass our het/homozygous calling\n",
    "mu = mu[(mu['IMPACT'].isin(['MODERATE', 'HIGH'])) & (mu['Call'].notnull()) & (mu['Call']!='unknown')]\n",
    "# Adding mutation ID column\n",
    "#mu['mut_ID'] = mu['GENE']+'_'+mu['EFFECT']+'_'mu['CHROM']+'_'+mu['POS'].apply(lambda p: str(p))+'_'+mu['ALT'].apply(lambda p: str(p))+'_'+mu['Call']\n",
    "#mu.to_csv('../Mutation_data/PLT_mutation_calls_simple.csv', index=False)\n",
    "\n",
    "#                                     #\n",
    "# Combining mutations near each other #\n",
    "#                                     #\n",
    "bc_to_mut_ids = defaultdict(list)\n",
    "# like bc_to_rows[bcs] = [[mut_id,mut_id], [mut_id], [mut_id,mut_id]] mut_ids combined in the inner lists if they are within 35 bp of another mut_id\n",
    "\n",
    "for jnk, row in mu.iterrows():\n",
    "    mid_new = (row['CHROM'], row['POS'], row['ALT'], row['Call'], row['GENE'], row['EFFECT'])\n",
    "    merged = []\n",
    "    mids = bc_to_mut_ids[row['Barcode']]\n",
    "    for i in range(len(mids)):\n",
    "        for mid_old in mids[i]:\n",
    "            if mid_old[0] == mid_new[0] and np.abs(int(mid_old[1])-int(mid_new[1]))<=35:\n",
    "                merged.append(i)\n",
    "    all_mids = []\n",
    "    mid_to_add = [mid_new]\n",
    "    for i in range(len(mids)):\n",
    "        if i in merged:\n",
    "            mid_to_add += mids[i]\n",
    "        else:\n",
    "            all_mids.append(mids[i])\n",
    "    all_mids.append(mid_to_add)\n",
    "    bc_to_mut_ids[row['Barcode']] = all_mids\n",
    "\n",
    "# Counting up how many times we see each gene hit\n",
    "gene_counts = Counter()\n",
    "for bc in bc_to_mut_ids:\n",
    "    for mid_group in bc_to_mut_ids[bc]:\n",
    "        gene_counts[mid_group[0][4]] += 1\n",
    "    \n",
    "#\n",
    "#  Creating a dataframe like BC, Gene_list (';'-separated), Effect_list (';'-sep), mut_ID_list (';'-sep, like CHROM_POS_ALT_Call)\n",
    "#  Genes, Effects, mut_IDs sorted by # hit (gene_counts)\n",
    "#  For merged mutations (within 35 bp of each other), Effects and mut_IDs are separated by spaces. For genes, we assume they are shared\n",
    "#\n",
    "mat = []\n",
    "for bc in bc_to_mut_ids:\n",
    "    tmprec = bc_to_mut_ids[bc]\n",
    "    tmp = [bc]\n",
    "    gene_list = []\n",
    "    #effect_list = []\n",
    "    mut_id_list = []\n",
    "    for entry_list in sorted(tmprec, key=lambda x: -1*gene_counts[x[0][4]]):\n",
    "        entry_list = list(set(entry_list)) # This deduplicates the same exact mutations found in multiple files associated with the same barcode\n",
    "        gene_list.append(entry_list[0][4])\n",
    "        #effect_list.append(' '.join([m[5] for m in entry_list]))\n",
    "        mut_id_list.append(' '.join([m[4]+'_'+m[0]+'_'+str(m[1])+'_'+m[2]+'_'+m[3]+'_'+m[5] for m in entry_list]))\n",
    "    \n",
    "    mat.append([bc, ';'.join(gene_list), ';'.join(mut_id_list)])\n",
    "\n",
    "\n",
    "md = pd.DataFrame(mat, columns=['Barcode', 'Genes', 'mut_IDs']) \n",
    "\n",
    "def putative_drivers(gene_list):\n",
    "    if len(gene_list) == 1:\n",
    "        return gene_list[0]\n",
    "    else:\n",
    "        drivers = [g for g in gene_list if gene_counts[g]>1]\n",
    "        return ';'.join(drivers)\n",
    "    \n",
    "def get_full_driver_mutations(row):\n",
    "    drivers = row['Putative_Drivers'].split(';')\n",
    "    genes = row['Genes'].split(';')\n",
    "    mutids = row['mut_IDs'].split(';')\n",
    "    return ';'.join(sorted([mutids[i] for i in range(len(genes)) if genes[i] in drivers]))\n",
    "    \n",
    "md['Putative_Drivers'] = md['Genes'].apply(lambda g: putative_drivers(g.split(';')))\n",
    "md['Putative_Driver_Mutations'] = md.apply(get_full_driver_mutations, axis=1)\n",
    "md.to_csv('Final_data_sets/Barcode_mutation_association.csv', index=False)\n",
    "md.head(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Methods\n",
    "* Call heterozygous or homozygous\n",
    "\n",
    "```\n",
    "if alt < 2:\n",
    "    return np.nan\n",
    "elif tot < 4:\n",
    "    return np.nan\n",
    "else:\n",
    "    if alt/tot >= 0.15 and alt/tot <= 0.85:\n",
    "        return 'Heterozygous'\n",
    "    elif alt/tot < 0.25:\n",
    "        return np.nan\n",
    "    else:\n",
    "        return 'Homozygous'\n",
    "```\n",
    "\n",
    "* Filter out synonymous / noncoding mutations and mutations that didn't pass our het/homozygous calling\n",
    "    * ` mu = mu[(mu['IMPACT'].isin(['MODERATE', 'HIGH'])) & (mu['Call'].notnull()) & (mu['Call']!='unknown')] `\n",
    "* Mutations get merged if they are within 35 bp of each other and associated with the same barcode (all mutations info from merged mutations will be separated by spaces)  \n",
    "* Call a mutations a putative driver if they are in a gene in which we observe at least two unique nonsyn mutations (very lenient definition)\n",
    "* Now, combine it with BFA data..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## OK Let's combine it with the BFA data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_identifier(row):\n",
    "    if pd.notnull(row['Putative_Driver_Mutations']):\n",
    "        return row['Putative_Driver_Mutations']\n",
    "    elif pd.notnull(row['mut_IDs']):\n",
    "        return row['mut_IDs']\n",
    "    else:\n",
    "        return row['Barcode']\n",
    "\n",
    "    \n",
    "fd = pd.read_csv('Final_data_sets/All_fitness_tidy.csv')\n",
    "bfa_dats = dict()\n",
    "good_envs = dict()\n",
    "envs_use = ['YPD', 'SC', '37C', 'pH3_8', 'pH7_3', 'GlyEtOH', 'FLC4', 'CLM', '21C', '02M_NaCl']\n",
    "icols = ['Barcode', 'Home_Environment', 'Putative_Neutral']\n",
    "for bfa in ['hBFA1_alpha', 'hBFA2_alpha', 'hBFA2_2N', 'dBFA2_2N']:\n",
    "    td = fd[(fd['BFA']==bfa.split('_')[0]) & (fd['ploidy']==bfa.split('_')[1])]\n",
    "    bfa_dats[bfa] = td[['Barcode', 'Test_Environment', 's']].pivot(index='Barcode', columns='Test_Environment', values='s')\n",
    "    good_envs[bfa] = [i for i in envs_use if len([j for j in bfa_dats[bfa][i] if pd.isnull(j)])!=len(bfa_dats[bfa])]\n",
    "    bfa_dats[bfa] = bfa_dats[bfa].merge(td[icols].drop_duplicates(), on='Barcode', how='left')[icols+good_envs[bfa]]\n",
    "\n",
    "# Adding mutation data\n",
    "# md = pd.read_csv('Final_data_sets/Barcode_mutation_association.csv')\n",
    "mut_based = dict()\n",
    "for bfa in bfa_dats:\n",
    "    bfa_dats[bfa] = bfa_dats[bfa].merge(md, on='Barcode', how='left')\n",
    "    bfa_dats[bfa]['Lineage_ID'] = bfa_dats[bfa].apply(get_identifier, axis=1)\n",
    "    # combining data by driver mutation(s), mutations, or barcode\n",
    "    mut_based[bfa] = bfa_dats[bfa][['Home_Environment', 'Lineage_ID']+good_envs[bfa]].groupby(['Home_Environment', 'Lineage_ID']).mean().reset_index()\n",
    "    mut_based[bfa] = mut_based[bfa].merge(bfa_dats[bfa][['Lineage_ID', 'Putative_Drivers']].drop_duplicates(), on='Lineage_ID', how='left')\n",
    "    mut_based[bfa]['Has_Mutation_Data'] = mut_based[bfa]['Lineage_ID'].apply(lambda x: '_' in x)\n",
    "    mut_based[bfa].to_csv('Final_data_sets/'+bfa+'_with_mutation_data.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hBFA1_alpha :   1810 lineages, 101 with mutation data. Envs:\n",
      "GlyEtOH_alpha    40\n",
      "CLM_alpha        33\n",
      "FLC4_alpha       28\n",
      "Name: Home_Environment, dtype: int64\n",
      "hBFA2_alpha :   1528 lineages, 41 with mutation data. Envs:\n",
      "CLM_alpha        33\n",
      "GlyEtOH_alpha     6\n",
      "FLC4_alpha        2\n",
      "Name: Home_Environment, dtype: int64\n",
      "hBFA2_2N :   840 lineages, 42 with mutation data. Envs:\n",
      "GlyEtOH_2N    27\n",
      "FLC4_2N       15\n",
      "Name: Home_Environment, dtype: int64\n",
      "dBFA2_2N :   4097 lineages, 88 with mutation data. Envs:\n",
      "CLM_2N         21\n",
      "21C_2N         17\n",
      "FLC4_2N        15\n",
      "GlyEtOH_2N     10\n",
      "37C_2N         10\n",
      "02M_NaCl_2N     9\n",
      "pH7_3_2N        6\n",
      "Name: Home_Environment, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# Showing which envs. have mutation data in each bfa\n",
    "for bfa in mut_based:\n",
    "    print(bfa, ':  ', len(mut_based[bfa]), 'lineages,', len(mut_based[bfa][mut_based[bfa]['Has_Mutation_Data']]), 'with mutation data. Envs:')\n",
    "    print(mut_based[bfa][mut_based[bfa]['Has_Mutation_Data']]['Home_Environment'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "milo_py37",
   "language": "python",
   "name": "milo_py37"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
